static/js/message_list_view.js

```diff
@@ -1445,4 +1445,76 @@ export class MessageListView {
             message_container.status_message = false;
         }
     }
+
+    update_sticky_recipient_headers() {
+        const rows_length = this._rows.size;
+        if (!rows_length) {
+            /* No headers are present */
+            return;
+        }
+        /* Intentionally remove sticky headers class here to make calculations simpler. */
+        $(".sticky_header").removeClass("sticky_header");
+        /* visible_top is navbar top position + height for us. */
+        const visible_top = $message_viewport.message_viewport_info().visible_top;
+        /* We need date to be properly visible on the header, so partially visible headers
+        who are about to be scrolled out of view are not acceptable. */
+        const $table = rows.get_table(this.table_name);
+        const $headers = $table.find(".message_header");
+        const header_height = $headers[0].getBoundingClientRect().height;
+
+        /* There are 3 types elements we expect to be present at this pixel:
+            * message header
+            * Date separator
+            * Z logo at the top of message feed
+
+            If we encounter either loading indicator or date separator, we set sticky header to the next recipient row. */
+        const elements_below_navbar = document.elementsFromPoint(
+            window.innerWidth / 2,
+            visible_top + header_height - 1,
+        );
+        let $sticky_header = $(
+            elements_below_navbar.filter((element) => element.classList.contains("message_header")),
+        );
+        const message_row = elements_below_navbar.filter((element) =>
+            element.classList.contains("message_row"),
+        );
+        const date_row = elements_below_navbar.filter((element) =>
+            element.classList.contains("date_row"),
+        );
+
+        /* Set correct date for the sticky_header. */
+        let $message_row;
+        if (date_row.length) {
+            /* There is always a message header present next to a date_row. */
+            const recipient_row = date_row[0].nextElementSibling;
```

Comment:
I don't like hardcoding assumptions about what elements we might have in the DOM -- could we instead do something like `.nextAll(".recipient_row").first()`?

Reply:
An alternative option, which might be faster, might be to just assert that we got a `.recipient_row` in a way that'll throw an exception in development? That will probably mean if we violate this invariant in a future feature, we'll notice.

Reply:
No longer relevant.